_published/0013-authentication.md

```diff
@@ -0,0 +1,165 @@
+
+#### Local development vs production environments
+For the initial release of the feature the configuration will be tailored to
+using azure as an IDP. These settings will as necessary be exposed as
+environment variables to each respective docker image.
+
+
+#### Choice of Client
+
+The OIDC standard defines a
+[list of certified clients](https://openid.net/developers/certified/)
+that can help simplify the implementation of ID Token retrieval.
+
+For the backend a full client will not be required. As the tokens just need
+to be validated and not requested. Validation can be done offline by caching
+the issuers' signature for future validations (OBS! The signature key may
+rotate on a timely basis).
+
+## Compatibility
+
+Expect no compatability between versions of wharf without authentication to
+versions of wharf with authentication. There is a potential threat in leaving
+the system backwards compatible in that a non-secured version of the api could
+run together in an otherwise secure cluster.
+
+We should aim to pick a solution where any IDP supporting OIDC and RBAC can be
+used. Locking ourselves to a single IDP when a standard such as OIDC exists
+seems like a  waste opportunity. Keeping the choice of IDP open will leave
+more room for choice and flexibility in the future.
+
+## Alternative solutions
+
+There are many types of identity verification systems.
+This RFC assumes that OIDC will be used. Within OIDC there are multiple
+different [flow definitions](https://developer.okta.com/docs/concepts/oauth-openid/#choosing-an-oauth-2-0-flow).
+We should primarily use `codeflow with pkce` as this is presumed to be the
+most secure, and best compatible at the time of writing.
+
+## Future possibilities
+
+We may want to implement multiple different IDP sources. This could allow
+login from different sources than the Iver-Azure-AD. A sample situation
+where this could be useful is where/when we wish to give a client
+view/control over the deployment pipeline.
+
+When handling ID-Tokens from RBAC enabled IDPs its is possible to verify
+custom permission levels using custom role claims.
+[See this link](https://docs.microsoft.com/en-us/azure/role-based-access-control/custom-roles)
+for more information about custom claims. This system allows for setup
+fine-tuned access control on a per-deployment basis. The initial setup
+for each of the different RBACs will likely require setup scripts on a
+per-provider basis as it requires interaction directly with each respective
+cloud on a non-standard basis.
+
+## Unresolved questions
+
+OIDC in any form demands full SSL security in all connections. In order to
+avoid having to fill the development environment with --force-unsecure
+bindings we should consider implementing certificate generation and importing
+for dev environments.
+
+Should we keep the wharf-api stateless and send the id-token in every
+request or should we create sessions and store the id token in each user's
+session?
```

Comment:
Actually, since it's a SPA web, we could just store it in memory and force a new ID-token on every page refresh. That way there's a smaller attack surface as no malicious code could find the token from cookies or localStorage.

Bonus point: if someone gets new roles in the IDP, all they have to do is refresh the page to have it apply

Reply:
> Actually, since it's a SPA web, we could just store it in memory and force a new ID-token on every page refresh. That way there's a smaller attack surface as no malicious code could find the token from cookies or localStorage.
> Bonus point: if someone gets new roles in the IDP, all they have to do is refresh the page to have it apply

This is mostly outside of our control. This should be correctly handled by the client oidc framework, and should automagically do more or less what you want but through different mechanics. I would reffer to look at the mechanics of the refresh token in the oidc standard.

> Is the idea not to store the ID token on the client-side? Like in localStorage or in cookies?

Yes but the question here is should we for localStorage or rely on sessionStorage. Other choices seem to be much more difficult as these are the two that are obviously supported by the OIDC client framework. One difference is how susceptible the sessions are to cors attacks. It should result in a bit of a different user experience between the two in terms of how well a user can stay logged in between different types of shutdown. After my last discussion with a colleague im leaning towards retaning the default setting of sessionStorage for increased protection against CORS attacks.

> Are ID-Tokens meant to only be seen by the server, or are they given to the server by the user? In the case of the latter, we could instead just rely on the user always providing the ID-token themselves on each request, letting them keep track of the session and keeping the wharf-api stateless.

Typically ID tokens are not supposed to be seen by the server as the whole point of the original design spec of OIDC is to allow a familiar authentication/authorisation flow without having to fully disclose ones identity. However with the introduction of RBACs, and their use in custom claims its possible to setup ID tokens in such a way that they dont GDPR relevant userdata while still being fully functional as authorization level proof much like an access code would in the OAUTH spec.

Reply:
My point there about only storing it in memory was that if we store the tokens in "memory", i.e. local/scoped javascript variables, instead of localStorage/sessionStorage then that makes it protected against XSS attacks, and since it's an SPA then the client won't need to make that many requests to the remote IDP anyway but instead only when opening a new tab or refreshing the page.

> Typically ID tokens are not supposed to be seen by the server /.../ its possible to setup ID tokens in such a way that they dont GDPR relevant userdata while still being fully functional as authorization level proof much like an access code would in the OAUTH spec.

That sh*t's baller :call_me_hand:

Reply:
JS variables are not a simple alternative here there may be a custom way of implementing it but since it would require alot more work i would see the present as a choice between local and session storage. Per meeting we belive sessionstorage is better than localstorage because the user can in a simpler way wipe the cache (opening/closing tab).